examples/modbus-dashboard/net.c

```diff
@@ -0,0 +1,256 @@
+  memset(&settings, 0, sizeof(settings));
+  mg_json_get_bool(body, "$.mqtt_enabled", &settings.mqtt_enabled);
+  settings.log_level = mg_json_get_long(body, "$.log_level", MG_LL_INFO);
+  setfromjson(body, "$.mqtt_server_url", &settings.mqtt_server_url);
+  setfromjson(body, "$.mqtt_topic_rx", &settings.mqtt_topic_rx);
+  setfromjson(body, "$.mqtt_topic_tx", &settings.mqtt_topic_tx);
+
+  s_settings = settings;  // TODO: save to the device flash
+  bool ok = true;
+  mg_http_reply(c, 200, s_json_header,
+                "{%m:%s,%m:%m}",                          //
+                MG_ESC("status"), ok ? "true" : "false",  //
+                MG_ESC("message"), MG_ESC(ok ? "Success" : "Failed"));
+}
+
+static void handle_settings_get(struct mg_connection *c) {
+  mg_http_reply(c, 200, s_json_header,
+                "{%m:%s,%m:%d,%m:%m,%m:%m,%m:%m}\n",  //
+                MG_ESC("mqtt_enabled"),
+                s_settings.mqtt_enabled ? "true" : "false",  //
+                MG_ESC("log_level"), s_settings.log_level,   //
+                MG_ESC("mqtt_server_url"), MG_ESC(s_settings.mqtt_server_url),
+                MG_ESC("mqtt_topic_rx"), MG_ESC(s_settings.mqtt_topic_rx),
+                MG_ESC("mqtt_topic_tx"), MG_ESC(s_settings.mqtt_topic_tx));
+}
+
+// Modbus handler function
+static void mfn(struct mg_connection *c, int ev, void *ev_data, void *fn_data) {
+  struct conndata *cd = (struct conndata *) c->data;
+  if (ev == MG_EV_READ) {
+    MG_INFO(("%lu RECEIVED %lu", c->id, c->recv.len));
+    if (c->recv.len < 8) return;  // Less than minimum length, buffer more
+    uint16_t len = mg_ntohs(*(uint16_t *) &c->recv.buf[4]);  // PDU length
+    if (c->recv.len < len + 4U) return;  // Partial frame, buffer more
+    // Notify parent connection
+    for (struct mg_connection *t = c->mgr->conns; t != NULL; t = t->next) {
+      if (t->id == cd->id) mg_call(t, MG_EV_USER, &c->recv);
+    }
+    c->is_closing = 1;
+  } else if (MG_EV_POLL) {
+    // MG_INFO(("%lu closing tmout %llu", c->id, cd->expiration_time));
+    if (cd->expiration_time > 0 && cd->expiration_time < mg_millis()) {
+      c->is_closing = 1;
+    }
+  }
+  (void) ev_data, (void) fn_data;
+}
+
+static struct mg_connection *start_modbus_request(struct mg_mgr *mgr,
+                                                  struct mg_str json,
+                                                  unsigned long cid) {
+  struct mg_connection *c = NULL;
+  char *url = mg_json_get_str(json, "$.url");
+  long timeout = mg_json_get_long(json, "$.timeout", 750);
+  uint8_t id = (uint8_t) mg_json_get_long(json, "$.id", 1);
+  uint16_t reg = mg_htons((uint16_t) (mg_json_get_long(json, "$.reg", 1) - 1));
+  // uint16_t val = (uint16_t) mg_json_get_long(json, "$.val", 0);
+  uint8_t func = (uint8_t) mg_json_get_long(json, "$.func", 0);
+  uint16_t nregs = mg_htons((uint16_t) mg_json_get_long(json, "$.nregs", 1));
+  MG_INFO(("%lu REQUEST: %.*s", cid, json.len, json.ptr));
```

Comment:
I think it is clearer to have 16-bit quantities in host order and only convert them to network order when sending.